gt/main.c

```diff
@@ -870,14 +794,310 @@ print_unsent_policy(struct ggu_policy *policy)
+		if (instance->ggu_pkts[i].buf == NULL)
+			continue;
+		send_notify_pkt(instance, &instance->ggu_pkts[i]);
+	}
+
+	instance->num_ggu_pkts = 0;
+}
+
+/*
+ * Start building a new notification packet for the Gatekeeper
+ * server indicated by @pkt_info.
+ *
+ * If there's no more room for a notification packet, then
+ * send a random one to make room.
+ */
+static struct ggu_notify_pkt *
+add_notify_pkt(struct gt_instance *instance,
+	struct gt_packet_headers *pkt_info)
+{
+	struct ggu_notify_pkt *ggu_pkt;
+	unsigned int lcore_id = rte_lcore_id();
+	unsigned int i;
+
+	if (instance->num_ggu_pkts == instance->max_ggu_pkts)
+		send_rand_notify_pkt(instance);
+	RTE_VERIFY(instance->num_ggu_pkts < instance->max_ggu_pkts);
+
+	for (i = 0; i < instance->max_ggu_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
+			break;
+	}
+	RTE_VERIFY(i < instance->max_ggu_pkts);
+
+	ggu_pkt = &instance->ggu_pkts[i];
+
+	ggu_pkt->ipaddr.proto = pkt_info->outer_ethertype;
+	if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *ipv4_hdr =
+			(struct ipv4_hdr *)pkt_info->outer_l3_hdr;
+		ggu_pkt->ipaddr.ip.v4.s_addr = ipv4_hdr->src_addr;
+	} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *ipv6_hdr =
+			(struct ipv6_hdr *)pkt_info->outer_l3_hdr;
+		rte_memcpy(ggu_pkt->ipaddr.ip.v6.s6_addr, ipv6_hdr->src_addr,
+			sizeof(ggu_pkt->ipaddr.ip.v6.s6_addr));
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u adding to notification packet to Gatekeeper server with unknown IP version %hu\n",
+			lcore_id, ggu_pkt->ipaddr.proto);
+	}
+
+	ggu_pkt->buf = rte_pktmbuf_alloc(
+		get_net_conf()->gatekeeper_pktmbuf_pool[
+			rte_lcore_to_socket_id(lcore_id)]);
+	if (ggu_pkt->buf == NULL) {
+		RTE_LOG(ERR, MEMPOOL,
+			"gt: failed to allocate notification packet on lcore %u!",
+			lcore_id);
+		return NULL;
+	}
+
+	instance->num_ggu_pkts++;
+	return ggu_pkt;
+}
+
+/*
+ * To estimate the maximum size of an on-the-wire policy decision,
+ * sum the size of the decision prefix (type and length fields) with
+ * the size of the in-memory GGU policy struct. This is a slight
+ * overestimate, which is acceptable for determining whether a
+ * packet has enough room for another decision.
+ */
+#define GGU_MAX_DECISION_LEN (sizeof(struct ggu_decision) + \
+	sizeof(struct ggu_policy))
+
+/*
+ * Add a policy decision to a notification packet. If a notification
+ * does not exist for this Gatekeeper server, then create one.
+ */
+static void
+fill_notify_pkt(struct ggu_policy *policy,
+	struct gt_packet_headers *pkt_info, struct gt_instance *instance,
+	struct gt_config *gt_conf)
+{
+	struct ggu_notify_pkt *ggu_pkt;
+	struct ggu_decision *ggu_decision;
+	ggu_pkt = find_notify_pkt(pkt_info, instance);
+	if (ggu_pkt == NULL) {
+		ggu_pkt = add_notify_pkt(instance, pkt_info);
+		if (ggu_pkt == NULL) {
+			print_unsent_policy(policy, NULL);
+			return;
+		}
+		fill_notify_pkt_hdr(ggu_pkt->buf, pkt_info, gt_conf);
```

Comment:
I think `add_notify_pkt()` should call `fill_notify_pkt_hdr()`, so the returned `ggu_pkt` is guaranteedly consistent.